Data contract nwb packaging

arjunsridhar12345 opened the pull request:
This PR attempts to create a primary nwb for the VR Foraging task by traversing through the data contract

arjunsridhar12345 committed:
Added environment.json

arjunsridhar12345 committed:
initial commit, linting

arjunsridhar12345 committed:
fix writing data types from software events to nwb

arjunsridhar12345 committed:
Initial readme description

arjunsridhar12345 committed:
update data contract, fix event description, linting. https://github.com/AllenNeuralDynamics/aind-physio-arch/issues/699

arjunsridhar12345 committed:
update description of meanings table

arjunsridhar12345 committed:
remove hardcoded check. remove extra variable for stream name. https://github.com/AllenNeuralDynamics/aind-physio-arch/issues/699

arjunsridhar12345 committed:
fix filename for torque stream

arjunsridhar12345 committed:
update readme for getting json data

arjunsridhar12345 committed:
update acquisition fields in readme

arjunsridhar12345 committed:
fix typo in readme

arjunsridhar12345 committed:
minor update to readme to make more general

arjunsridhar12345 committed:
add try except for getting stream

arjunsridhar12345 committed:
remove unused classes for attempting hierarchy. simplify functions and main run

arjunsridhar12345 committed:
fix json reading

arjunsridhar12345 committed:
simplify logic, remove unused functions

arjunsridhar12345 committed:
add capsule link to readme

arielleleon commented on code/run_capsule.py:
```diff
@@ -1,7 +1,132 @@
-""" top level run script """
+import json
+import logging
+from datetime import datetime
```
import not used

arielleleon commented on code/run_capsule.py:
```diff
@@ -1,7 +1,132 @@
+        level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
+    )
+
+    primary_data_path = tuple(utils.DATA_PATH.glob("*"))
```
I would not use a hardcoded path here. Users should have the option to add in their own input and output directories

arielleleon replied:
I would always code with that in mind. I think `argparse` or `BaseSettings` will work

arielleleon commented on code/run_capsule.py:
```diff
@@ -1,7 +1,132 @@
+    if not primary_data_path:
+        raise FileNotFoundError("No primary data asset attached")
+
+    if len(primary_data_path) > 1:
```
What happens when there are multiple modalities in the data path?

arjunsridhar12345 replied:
I think if I understand this, multiple modalities are ok. That would depend on the data contract specifying 
` root path / modality / ...`. If multiple data assets are attached (there are then multiple root paths) so not sure which to use? I think the variable name is confusing - will rename it to data_root_path or something similar

arielleleon replied:
The problem is the entire primary asset will be attached. If you glob `/data` then you will return all the directories and files associated with the primary asset.

arielleleon commented on code/run_capsule.py:
```diff
@@ -1,7 +1,132 @@
+                logger.info(f"Failed to load {stream.name} with exception {e}")
+        elif isinstance(stream, data_contract.SoftwareEvents):
+            try:
+                data = utils.clean_dataframe_for_nwb(stream.data.reset_index())
```
Does this add NaN's to the event df?

arjunsridhar12345 replied:
it should replace any `NoneType` cells with NaN's for nwb compliant types. So, I guess yes, hehe it is adding NaN's. That function is just what data types I found needed to be convert to a nwb compliant data type.

arielleleon replied:
Okay, maybe we can talk briefly over the nwb tomorrow

arielleleon reviewed (changes requested):
Can you run flake8? Few small changes requested. I would also like to look at the NWB output with you before approval. Thanks!

arjunsridhar12345 committed:
start addressing PR feedback. update data contract dependency. https://github.com/AllenNeuralDynamics/aind-vr-foraging-primary-data-nwb-packaging/pull/1

arjunsridhar12345 committed:
update readme with specific keys that have json data

arjunsridhar12345 committed:
remove hardcoded paths and use pydantic base settings. 
https://github.com/AllenNeuralDynamics/aind-vr-foraging-primary-data-nwb-packaging/pull/1

arjunsridhar12345 committed:
add link to data contract and minor restructure

arjunsridhar12345 committed:
use generic mount name

arjunsridhar12345 committed:
use camel case for settings class name

arjunsridhar12345 committed:
use uv to bootstrap dependencies

arjunsridhar12345 committed:
simplify output name of raw nwb. remove adding events to raw nwb for now

arjunsridhar12345 committed:
add pulling version from tasklogic file. https://github.com/AllenNeuralDynamics/aind-physio-arch/issues/774. fix linting

arjunsridhar12345 commented:
This fell off, and there has been some changes. Apologies on my end but summary below:

- Data contract now in Aind.Behavior.VRForaging
- Uses version from tasklogic input to pass to data contract
- Uses uv for dependencies (Bruno did this for the qc capsule). It is very fast and easier to manage but can change if needed

bruno-f-cruz commented on code/run_capsule.py:
```diff
@@ -1,7 +1,150 @@
+    vr_foraging_dataset = dataset(
+        primary_data_path[0], version=contract_version
+    )
+    exec = vr_foraging_dataset.load_all()  # load tree structure
```
If you are not loading behavior-videos, you can do `vr_foraging["Behavior"].load_all()`

bruno-f-cruz commented on code/run_capsule.py:
```diff
@@ -1,7 +1,150 @@
+    event_data = []  # for adding to events table
+
+    # using this ndx object for events table
+    nwb_file = NdxEventsNWBFile(
```
Small note, `NdxEventsNWBFile` should be `NdxEventsNwbFile`

bruno-f-cruz replied:
Ok ignore this, It seems NWB/Zarr/HDF5 doesn't care about Pascal case conventions. `NWBZarrIO` has been deployed already. For the future, I really advise people to stick with this standard: https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions#capitalization-rules-for-identifiers if at all possible.

arjunsridhar12345 replied:
ok I was trying to follow this: https://github.com/rly/ndx-events/blob/main/examples/write_ttls_events.py. there its `NdxEventsNWBFile`

bruno-f-cruz commented on code/run_capsule.py:
```diff
@@ -1,7 +1,150 @@
+        name = stream.resolved_name.replace("::", ".")
+        name = name[name.index(".") + 1:]
+        if isinstance(
+            stream.parent, data_contract.harp.HarpDevice
```
You can do `data_contract.harp.HarpRegister` directly I think

bruno-f-cruz commented on code/utils.py:
```diff
@@ -0,0 +1,82 @@
+    return data
+
+
+def datetime_to_iso_in_dict(
```
Can you explain why this is needed? I believe I always use iso in everything, but i may be wrong

bruno-f-cruz commented on environment/pyproject.toml:
```diff
@@ -0,0 +1,20 @@
+readme = "README.md"
+requires-python = ">=3.12"
+dependencies = [
+    "aind-behavior-vr-foraging[data]",
```
This is on me, but for now please lock this to a specific commit. I will make a release soon and you can lock to it then.

bruno-f-cruz commented on environment/pyproject.toml:
```diff
@@ -0,0 +1,20 @@
+requires-python = ">=3.12"
+dependencies = [
+    "aind-behavior-vr-foraging[data]",
+    "aind-data-schema>=1.4.0",
```
This is locked via aind-behavior-vr-foraging. You can use the `[dev]` as a way to get all these requirements.

bruno-f-cruz commented on environment/pyproject.toml:
```diff
@@ -0,0 +1,20 @@
+    "aind-behavior-vr-foraging[data]",
+    "aind-data-schema>=1.4.0",
+    "aind-data-schema-models>=0.7.5",
+    "contraqctor>=0.4.3",
```
See previous comment

bruno-f-cruz commented on environment/pyproject.toml:
```diff
@@ -0,0 +1,20 @@
+
+[tool.uv.sources]
+aind-behavior-vr-foraging = { git = "https://github.com/AllenNeuralDynamics/Aind.Behavior.VrForaging.git"}
+ndx-events = { git = "https://github.com/rly/ndx-events.git"}
```
I thought this extension was merged to main in the NWB

arjunsridhar12345 replied:
I think the latest version hasn't been released yet: https://github.com/rly/ndx-events. Also I've never really followed how extensions actually work in NWB

bruno-f-cruz reviewed (changes requested):
Mostly good. I left a few small notes to try to simplify dependencies.

arielleleon reviewed (approved):
Looks great! This is a good capsule to model the harp behavior nwb packaging template from

arjunsridhar12345 committed:
revert run and docker file

arjunsridhar12345 committed:
put back uv dependency

arjunsridhar12345 committed:
put back uv in run.sh

arjunsridhar12345 committed:
refactor after PR feedback.

arjunsridhar12345 committed:
add subject to nwb

arjunsridhar12345 committed:
docs: update readme
